### DISCUSSION TITLE
Real-time change to Follow Group Array

### DISCUSSION THREAD

**Discussion by mousebuttons** [ramokz/phantom-camera#141] [Question / Help Wanted] [answered]
Hello Phantom Camera discussions!

I'm using PhantomCamera2D to follow two players in a co-op game, but the players ("aloe" and "vera") are destroyed and re-instantiated all the time. Using a script, I erase the follow group array and refill it with the new character nodes. My script is below:

````
extends Camera2D

var pcam: PhantomCamera2D
var aloe
var vera

# Called when the node enters the scene tree for the first time.
func _ready():
	pcam = get_node("../PhantomCamera2D")
	aloe = get_node("../Aloe/Head")
	vera = get_node("../Vera")

func _process(delta):
	print(global_position)

func _on_game_manager_aloe_spawn():
	print(pcam.get_follow_group_nodes())
	pcam.erase_follow_group()
	aloe = get_node("../Aloe/Head")
	pcam.append_follow_group_node(aloe)
	pcam.append_follow_group_node(vera)
	print(pcam.get_follow_group_nodes())
````

From the print statements I can tell that by the time this function is called, the follow group contains a Freed Object (the Aloe character who gets destroyed). I wrote the `erase_follow_group()` function into phantom_camera_2D.gd because `erase_follow_group_node(value: Node2D` requires an existing node to remove, and a Freed Object doesn't count. Here is the function I wrote:

````
## Removes every Node2D from Follow Group array.
func erase_follow_group() -> void:
	Properties.follow_group_nodes_2D.clear()
	Properties.should_follow = false
	Properties.has_follow_group = false
````

These functions clear the array and fill it with the new nodes, and printing the camera's position shows it to be tracking both characters normally. But nothing draws except the background permanently after Aloe respawns and the array is overwritten. I feel like I'm going against the camera system's design with my functions, but it's the only way I can think of. How can I troubleshoot this, or what would you do to swap out the follow targets?

Thanks for any help you can offer!

**Comment by mousebuttons** [comment]
Update: I have some good news! If you use replace_by() on the node in the Follow Group, there is no need to swap out the array.

https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-method-replace-by

Hope this can help someone in the future :)

**Reply by ramokz** [reply]
This is a great find!

If I understand it correctly, are you essentially just calling `replace_by()` and removed the entire `erase_follow_group()` function call?

**Reply by mousebuttons** [reply]
Thanks!

Yes, if you call `replace_by()` on one of the nodes in the follow group, it seamlessly follows the new node (I'm very impressed and so glad this works).

`erase_follow_group()` is no longer used, although I still wonder why it didn't work. I may look into it just out of curiosity. I'll post again if things develop.

**Reply by ramokz** [reply]
Nice, I'll be sure to add your finding to the documentation page, can definitely imagine others also finding it useful.

Thanks for sharing this!

**Reply by mousebuttons** [reply]
Update: I did not test my solution enough. It doesn't work in all cases, and I may have been mistaken about it working at all. I will post again when I have more information about it.

**Comment by mousebuttons** [comment, marked as answer]
Hello again!

I am back to describe how I've made it work for real.
Going back to my example of Aloe and Vera,
```cs 
// Make the new Aloe character and add it to the scene.
Aloe temporaryNewAloe = (Aloe)aloeScene.Instantiate();
GetParent().AddChild(temporaryNewAloe);

// Remove old Aloe from the follow group, then queue it for deletion.
phantomCam.Call("erase_follow_group_node", aloe.GetNode<RigidBody2D>("Head"));
aloe.QueueFree();

// Set our Aloe reference to the new Aloe.
aloe = temporaryNewAloe;

// Put in the New Aloe in the follow group.
phantomCam.Call("append_follow_group_node", aloe.GetNode<RigidBody2D>("Head"));
```

So I just needed to base my design more around the intended design of the Phantom Cam functions. As long as the node isn't freed until the end of the frame, Phantom Camera gets to access it and remove it from the follow group. Adding the new node in can happen during the same frame.

I experimented more with `replace_by` but found this solution before I could make `replace_by` work. It might be worth investigating further but I know this works.

Also during this journey I found that `_interpolate_position` in `phantom_camera_2D` does not check for NaN in the `position` Vector2. This makes it possible for the camera position to be set to NaN if a follow target has a physics glitch. I wrote a fix for this and can put it in a pull request when I've cleaned up my code.

Thanks for your support! The codebase has been easy to read and that's helped a lot.
